tasks/Task.cpp

```diff
@@ -26,178 +26,128 @@ Task::~Task()
-    int width;
-    BVTMagImage_GetWidth(img, &width);
-    printf("BVTMagImage_GetWidth: %d\n", width);
+        //set TVG Slope
+        head.SetTVGSlope(_tvg_slope.get());
 
-    // Save it to a PGM (PortableGreyMap)
-    ret = BVTMagImage_SavePGM(img, "img.pgm");
-    if( ret != 0 )
-    {
-        printf("BVTMagImage_SavePGM: ret=%d\n", ret);
-        return 1;
-    }
+        //set the speed of sound
+        head.SetSoundSpeed(_sound_speed.get());
 
-    /////////////////////////////////////////////////////////
+        //set Fluid Type
+        this->setFluidType(_fluid_type.get());
 
-    // Build a color mapper
-    BVTColorMapper mapper;
-    mapper = BVTColorMapper_Create();
-    if( mapper == NULL )
-    {
-        printf("BVTColorMapper_Create: failed\n");
-        return 1;
-    }
+        //set image processing resolution
+        this->setImageResolution(_image_resolution.get());
 
-    // Load the bone colormap
-    ret = BVTColorMapper_Load(mapper, "bone.cmap");
-    if( ret != 0 )
-    {
-        printf("BVTColorMapper_Load: ret=%d\n", ret);
-        return 1;
-    }
+        /////// Create a file to save the pings    /////
+        if (_save_file.get())
+        {
+            file.CreateFile(_file_name.get().c_str(), son, "");
+            // Request the first head for the file
+            head_file = file.GetHead(0);
+        }
 
-
-    // Perform the colormapping
-    BVTColorImage cimg;
-    ret = BVTColorMapper_MapImage(mapper, img, &cimg);
-    if( ret != 0 )
-    {
-        printf("BVTColorMapper_MapImage: ret=%d\n", ret);
-        return 1;
+        first_ping = true;
     }
-    printf("\n");
-
-    /////////////////////////////////////////////////////////
-    // Check the image height and width out
-    BVTColorImage_GetHeight(cimg, &height);
-    printf("BVTColorImage_GetHeight: %d\n", height);
-    BVTColorImage_GetWidth(cimg, &width);
-    printf("BVTColorImage_GetWidth: %d\n", width);
-
-
-    // Save it to a PPM (PortablePixMap)
-    ret = BVTColorImage_SavePPM(cimg, "cimg.ppm");
-    if( ret != 0 )
+    catch (BVTSDK::SdkException &e)
     {
-        printf("BVTColorImage_SavePPM: ret=%d\n", ret);
-        return 1;
+        std::cout << e.ErrorName() << std::endl;
+        throw std::runtime_error(e.ErrorMessage());
     }
 
-    // Clean up
-    BVTColorImage_Destroy(cimg);
-    BVTMagImage_Destroy(img);
-    BVTColorMapper_Destroy(mapper);
-    BVTPing_Destroy(ping);
-    BVTSonar_Destroy(son);
     return true;
+    
 }
 bool Task::startHook()
 {
     if (! TaskBase::startHook())
+    {
         return false;
+    }
     return true;
 }
 void Task::updateHook()
 {
     TaskBase::updateHook();
+
+    try
+    {
+        //Get Ping
+        BVTSDK::Ping ping(head.GetPing(-1));
+
+        // Save ping in the file
+        if(_save_file.get())
+        {
+            head_file.PutPing(ping);
+        }
+
+        // Generate an image from the ping
+        BVTSDK::ImageGenerator img_gen;
```

Comment:
I would expect the image generator creation to be a pretty costly thing. Why do you create it in the hot path instead of doing it once in configureHook or startHook ?